### PR TITLE
Development

### LINKED ISSUE #10: Special characters
Non-standard characters (such as Mac quotes and probably emojis) are rejected by both the Worker and Comlink. A way must be found to handle a broader set of characters.
Example trace:
SyntaxError: invalid character '“' (U+201C)
comlink.ts:544Unhandled Promise Rejection: DataCloneError: The object can not be cloned.

After some research, it seems the DataCloneError only occurs on Safari. Chrome handles it just fine.

### PR DESCRIPTION
**Pull request by winniederidder**
Contains fixes for the following features:

Where possible, console.log and similar calls have been replaced by calls to papyrosLog, to reduce the amount of logs in production that might be useful in development.

The inputServiceWorker now adds extra headers to responses from requests for urls within our domain, allowing the use of SharedArrayBuffers even in GitHub pages. Locally this worked thanks to the webpack-devServer, but these headers are not added by GitHub. Now the more efficient input calls should work in supporting browsers. This logic has also been moved to index.ts, as it relies on a reload of the page. After starting the service worker, we need a reload to ensure that our page is fetched with the correct headers. A key in localStorage tells the app whether it can proceed or needs to reload.

The Python scoping is now correct. Defining a variable in a code execution run, will no longer cause it to be available at next runs, as it is now a separate namespace.

Errors thrown by running the code are converted into strings, which hopefully fixes #10 as errors on Safari seem to have a different shape, causing them to be uncloneable by Comlink.

### CODE REVIEW

**Review by bmesuere — CHANGES_REQUESTED** [review]
- Please use meaningful commits in the future. "Fix a lot of things" makes it harder to review.
- As explained on monday, I'm still not a fan of the async input solution. Since you merged it anyway, I suggest keeping it in but definitely don't expand this solution. For future reference, I'll list my concerns below.

- Using atomics for waiting is a hack. This forces us into using SharedArrayBuffers to pass data which requires encoding/decoding strings and additional workarounds to pass the size of the input.
- Atomics are not supported by all browsers Dodona supports. The fallback, using local urls, is also a hack and requires registering a service worker to set custom headers.
- Service workers are powerful, but very easy to mess up. Especially when you need to update them. If the website where papyros is used already uses service workers, this might cause a conflict.
- A lot of the current code complexity in papyros comes from the input handling. This code is hard to test and we have only minimal error reporting from client side scripts. If this breaks in some browser, we won't be notified automatically. I am not confident using this in production.
- This is something that should clearly be fixed upstream.

File: src/Backend.ts [hunk 1]
```diff
@@ -52,9 +54,23 @@ export abstract class Backend {
...
+        return this._runCodeInternal(code)
+            .then(data => {
+                papyrosLog(LogType.Important, "ran code: " + code + " and received: ", data);
+                return this.onEvent({type: "success", data: data});
+            })
+            .catch(error => {
+                const errorString = "toString" in error ? error.toString() : JSON.stringify(error);
+                papyrosLog(LogType.Error, "Error during execution: ", error, errorString);
+                return this.onEvent({type: "error", data: errorString});
+            });
```
[the remark this answers is not in the document]
Comment (winniederidder) [reply src/Backend.ts:74]: An issue with this is that async and abstract cannot be combined, so I'm unsure if I can cleanly change it here.

[the remark this answers is not in the document]
Comment (bmesuere) [reply src/Backend.ts:74]: You are right for the definition part, but the implementation of that function can be async. In addition, when implementing/extending functions, always add the ["override" keyword](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-3.html#override-and-the---noimplicitoverride-flag) for additional checks.

**Comment by winniederidder** [comment]
The build update commits should fix #14 .

**Review by bmesuere — APPROVED** [review]
I suggest we merge this in main and work with clean, small branches/PRs from now on.

**Pull request merged**